File: iosApp/flare/UI/Component/TabIcon.swift
import SwiftUI
import KotlinSharedUI

struct TabTitle: View {
    let title: Any
    var body: some View {
        Text(String(describing: title))
    }
}

struct TimelineTabTitle: View {
    let title: UiText
    var body: some View {
        Text(title.text)
    }
}

extension UiText {
    var text: String {
        switch onEnum(of: self) {
        case .localized(let localized):
            return localized.string.text
        case .raw(let raw):
            return raw.string
        }
    }
}

extension UiStrings {
    var text: String {
        switch self {
        case .home: String(localized: "home_tab_home_title")
        case .notifications: String(localized: "home_tab_notifications_title")
        case .discover: String(localized: "home_tab_discover_title")
        case .me: String(localized: "home_tab_me_title")
        case .settings: String(localized: "settings_title")
        case .mastodonLocal: String(localized: "mastodon_tab_local_title")
        case .mastodonPublic: String(localized: "mastodon_tab_public_title")
        case .featured: String(localized: "home_tab_featured_title")
        case .bookmark: String(localized: "home_tab_bookmarks_title")
        case .favourite: String(localized: "home_tab_favorite_title")
        case .list: String(localized: "home_tab_list_title")
        case .feeds: String(localized: "home_tab_feeds_title")
        case .directMessage: String(localized: "dm_list_title")
        case .rss: String(localized: "rss_title")
        case .social: String(localized: "social_title")
        case .antenna: String(localized: "antenna_title")
        case .mixedTimeline: String(localized: "mixed_timeline_title")
        case .liked: String(localized: "liked_tab_title")
        case .allRssFeeds: String(localized: "all_rss_feeds_title")
        case .posts: String(localized: "posts_title")
        case .postsWithReplies: String(localized: "posts_with_replies_title", defaultValue: "Posts & Replies")
        case .media: String(localized: "media_title", defaultValue: "Media")
        case .channel: String(localized: "channel_title")
        case .default: String(localized: "tab_settings_default")
        case .login: String(localized: "login_button")
        case .verify: String(localized: "verify_button", defaultValue: "Verify")
        case .cancel: String(localized: "cancel_button")
        case .next: String(localized: "service_select_next_button")
        case .username: String(localized: "bluesky_login_username_hint")
        case .password: String(localized: "bluesky_login_password_hint")
        case .otp: String(localized: "bluesky_login_auth_factor_token_hint")
        case .oauthLogin: String(localized: "bluesky_login_oauth_button")
        case .passwordLogin: String(localized: "bluesky_login_use_password_button")
        case .qrConnect: String(localized: "nostr_login_qr_button")
        case .credentialImport: String(localized: "nostr_login_title")
        case .externalSigner: String(localized: "nostr_login_amber_button")
        case .webCookieLogin: String(localized: "login_button")
        case .nostrLoginAccount: String(localized: "nostr_login_account_hint")
        case .following: String(localized: "misskey_channel_tab_following")
        case .pixivRankingWeek: String(localized: "pixiv_ranking_week_title", defaultValue: "Weekly Ranking")
        case .pixivRankingMonth: String(localized: "pixiv_ranking_month_title", defaultValue: "Monthly Ranking")
        case .pixivRankingDayMale: String(localized: "pixiv_ranking_day_male_title", defaultValue: "Male Ranking")
        case .pixivRankingDayFemale: String(localized: "pixiv_ranking_day_female_title", defaultValue: "Female Ranking")
        case .pixivRankingWeekOriginal: String(localized: "pixiv_ranking_week_original_title", defaultValue: "Original Ranking")
        case .pixivRankingWeekRookie: String(localized: "pixiv_ranking_week_rookie_title", defaultValue: "Rookie Ranking")
        case .pixivRankingDayManga: String(localized: "pixiv_ranking_day_manga_title", defaultValue: "Manga Ranking")
        case .illustrations: String(localized: "illustrations_title", defaultValue: "Illustrations")
        case .manga: String(localized: "manga_title", defaultValue: "Manga")
        }
    }
}

struct TabIcon: View {
    let icon: IconType
    let size: CGFloat
    let iconOnly: Bool

    init(
        icon: IconType,
        size: CGFloat = 20,
        iconOnly: Bool = false
    ) {
        self.icon = icon
        self.size = size
        self.iconOnly = iconOnly
    }

    var body: some View {
        switch onEnum(of: icon) {
        case .material(let material):
            MaterialTabIcon(icon: material.icon)
                .frame(width: size, height: size)
        case .avatar(let avatar):
            AvatarTabIcon(userKey: avatar.accountKey, accountType: AccountType.Specific(accountKey: avatar.accountKey))
                .frame(width: size, height: size)
        case .url(let url):
            NetworkImage(data: url.url)
                .frame(width: size, height: size)
        case .mixed(let mixed):
            if iconOnly {
                MaterialTabIcon(icon: mixed.icon)
                    .frame(width: size, height: size)
            } else {
                ZStack(
                    alignment: .bottomTrailing
                ) {
                    AvatarTabIcon(userKey: mixed.accountKey, accountType: AccountType.Specific(accountKey: mixed.accountKey))
                        .frame(width: size, height: size)
                    MaterialTabIcon(icon: mixed.icon)
                        .padding(2)
                        .background(Color.white)
                        .foregroundStyle(Color.black)
                        .clipShape(.circle)
                        .frame(width: size / 2, height: size / 2)
                }
                .frame(width: size, height: size)
            }
        case .favIcon(let favIcon):
            FavTabIcon(host: favIcon.host)
                .frame(width: size, height: size)
        }
    }
}

extension TabIcon {
    init(
        tabItem: UiTimelineTabItem,
        size: CGFloat = 20,
        iconOnly: Bool = false
    ) {
        self.init(icon: tabItem.icon, size: size, iconOnly: iconOnly)
    }
}

struct MaterialTabIcon: View {
    let icon: UiIcon
    var body: some View {
        Image(icon.imageName)
            .resizable()
            .scaledToFit()
    }
}

struct AvatarTabIcon: View {

    @StateObject private var presenter: KotlinPresenter<UserState>

    init(userKey: MicroBlogKey, accountType: AccountType) {
        self._presenter = .init(wrappedValue: .init(presenter: UserPresenter(accountType: accountType, userKey: userKey)))
    }

    var body: some View {
        StateView(state: presenter.state.user) { user in
            AvatarView(data: user.avatar?.url, customHeader: user.avatar?.customHeaders)
        } loadingContent: {
            Image("fa-globe")
                .resizable()
                .scaledToFit()
                .redacted(reason: .placeholder)
        }
    }
}

struct FavTabIcon: View {
    @StateObject private var presenter: KotlinPresenter<UiState<NSString>>
    
    init(host: String) {
        self._presenter = .init(wrappedValue: .init(presenter: FavIconPresenter(host: host)))
    }
    
    var body: some View {
        StateView(state: presenter.state) { url in
            NetworkImage(data: .init(url))
        } loadingContent: {
            Image("fa-globe")
                .resizable()
                .scaledToFit()
                .redacted(reason: .placeholder)
        }
    }
}

File: compose-ui/src/commonMain/kotlin/dev/dimension/flare/ui/component/TabIcon.kt
package dev.dimension.flare.ui.component

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import dev.dimension.flare.compose.ui.Res
import dev.dimension.flare.compose.ui.all_rss_feeds_title
import dev.dimension.flare.compose.ui.antenna_title
import dev.dimension.flare.compose.ui.channel_title
import dev.dimension.flare.compose.ui.dm_list_title
import dev.dimension.flare.compose.ui.home_tab_bookmarks_title
import dev.dimension.flare.compose.ui.home_tab_discover_title
import dev.dimension.flare.compose.ui.home_tab_favorite_title
import dev.dimension.flare.compose.ui.home_tab_featured_title
import dev.dimension.flare.compose.ui.home_tab_feeds_title
import dev.dimension.flare.compose.ui.home_tab_home_title
import dev.dimension.flare.compose.ui.home_tab_list_title
import dev.dimension.flare.compose.ui.home_tab_me_title
import dev.dimension.flare.compose.ui.home_tab_notifications_title
import dev.dimension.flare.compose.ui.liked_title
import dev.dimension.flare.compose.ui.mastodon_tab_local_title
import dev.dimension.flare.compose.ui.mastodon_tab_public_title
import dev.dimension.flare.compose.ui.mixed_timeline_title
import dev.dimension.flare.compose.ui.posts_title
import dev.dimension.flare.compose.ui.rss_title
import dev.dimension.flare.compose.ui.settings_title
import dev.dimension.flare.compose.ui.social_title
import dev.dimension.flare.data.model.IconType
import dev.dimension.flare.data.model.TabItem
import dev.dimension.flare.data.model.TitleType
import dev.dimension.flare.model.AccountType
import dev.dimension.flare.ui.component.platform.PlatformText
import dev.dimension.flare.ui.component.status.toImageVector
import dev.dimension.flare.ui.component.platform.PlatformTextStyle
import dev.dimension.flare.ui.model.onLoading
import dev.dimension.flare.ui.model.onSuccess
import dev.dimension.flare.ui.presenter.home.FavIconPresenter
import dev.dimension.flare.ui.presenter.home.UserPresenter
import dev.dimension.flare.ui.presenter.invoke
import dev.dimension.flare.ui.theme.PlatformContentColor
import dev.dimension.flare.ui.theme.PlatformTheme
import moe.tlaster.precompose.molecule.producePresenter
import org.jetbrains.compose.resources.StringResource
import org.jetbrains.compose.resources.stringResource
import kotlin.native.HiddenFromObjC

@HiddenFromObjC
@Composable
public fun TabTitle(
    title: TitleType,
    modifier: Modifier = Modifier,
    color: Color = Color.Unspecified,
    fontSize: TextUnit = TextUnit.Unspecified,
    style: TextStyle = PlatformTextStyle.current,
) {
    PlatformText(
        text =
            when (title) {
                is TitleType.Localized -> stringResource(title.res)
                is TitleType.Text -> title.content
            },
        modifier = modifier,
        color = color,
        fontSize = fontSize,
        style = style,
        maxLines = 1,
    )
}

@HiddenFromObjC
@Composable
public fun TabIcon(
    tabItem: TabItem,
    modifier: Modifier = Modifier,
    iconOnly: Boolean = false,
    color: Color = PlatformContentColor.current,
) {
    TabIcon(
        accountType = tabItem.account,
        icon = tabItem.metaData.icon,
        title = tabItem.metaData.title,
        modifier = modifier,
        iconOnly = iconOnly,
        color = color,
    )
}

@HiddenFromObjC
@Composable
public fun TabIcon(
    accountType: AccountType,
    icon: IconType,
    title: TitleType,
    modifier: Modifier = Modifier,
    iconOnly: Boolean = false,
    color: Color = PlatformContentColor.current,
    size: Dp = 24.dp,
) {
    when (icon) {
        is IconType.Avatar -> {
            val userState by producePresenter(key = "$accountType:${icon.userKey}") {
                remember(accountType, icon.userKey) {
                    UserPresenter(
                        accountType,
                        icon.userKey,
                    )
                }.invoke()
            }
            userState.user
                .onSuccess {
                    AvatarComponent(it.avatar, size = size, modifier = modifier)
                }.onLoading {
                    AvatarComponent(null, size = size, modifier = modifier.placeholder(true))
                }
        }

        is IconType.Material -> {
            FAIcon(
                imageVector = icon.icon.toImageVector(),
                contentDescription =
                    when (title) {
                        is TitleType.Localized -> stringResource(title.res)
                        is TitleType.Text -> title.content
                    },
                modifier =
                    modifier
                        .size(size),
                tint = color,
            )
        }

        is IconType.Mixed -> {
            if (iconOnly) {
                FAIcon(
                    imageVector = icon.icon.toImageVector(),
                    contentDescription =
                        when (title) {
                            is TitleType.Localized -> stringResource(title.res)
                            is TitleType.Text -> title.content
                        },
                    modifier =
                        modifier
                            .size(size),
                    tint = color,
                )
            } else {
                val userState by producePresenter(key = "$accountType:${icon.userKey}") {
                    remember(accountType, icon.userKey) {
                        UserPresenter(
                            accountType,
                            icon.userKey,
                        )
                    }.invoke()
                }
                Box(
                    modifier = modifier,
                ) {
                    userState.user
                        .onSuccess {
                            AvatarComponent(it.avatar, size = size)
                        }.onLoading {
                            AvatarComponent(
                                null,
                                size = size,
                                modifier = Modifier.placeholder(true),
                            )
                        }
                    FAIcon(
                        imageVector = icon.icon.toImageVector(),
                        contentDescription =
                            when (title) {
                                is TitleType.Localized -> stringResource(title.res)
                                is TitleType.Text -> title.content
                            },
                        modifier =
                            Modifier
                                .size(size / 2)
                                .align(Alignment.BottomEnd)
                                .background(PlatformTheme.colorScheme.card, shape = CircleShape)
                                .padding(2.dp),
                    )
                }
            }
        }

        is IconType.Url -> {
            NetworkImage(
                icon.url,
                contentDescription =
                    when (title) {
                        is TitleType.Localized -> stringResource(title.res)
                        is TitleType.Text -> title.content
                    },
                modifier =
                    modifier
                        .size(size),
                contentScale = ContentScale.Fit,
            )
        }

        is IconType.FavIcon -> {
            val iconState by producePresenter(key = "fav-$accountType:${icon.host}") {
                remember(accountType, icon) {
                    FavIconPresenter(
                        icon.host,
                    )
                }.invoke()
            }
            iconState
                .onSuccess {
                    NetworkImage(
                        it,
                        contentDescription =
                            when (title) {
                                is TitleType.Localized -> stringResource(title.res)
                                is TitleType.Text -> title.content
                            },
                        modifier =
                            modifier
                                .size(size),
                        contentScale = ContentScale.Fit,
                    )
                }.onLoading {
                    AvatarComponent(
                        null,
                        size = size,
                        modifier = Modifier.placeholder(true),
                    )
                }
        }
    }
}

internal val TitleType.Localized.res: StringResource
    get() =
        when (key) {
            TitleType.Localized.LocalizedKey.Home -> Res.string.home_tab_home_title
            TitleType.Localized.LocalizedKey.Notifications -> Res.string.home_tab_notifications_title
            TitleType.Localized.LocalizedKey.Discover -> Res.string.home_tab_discover_title
            TitleType.Localized.LocalizedKey.Me -> Res.string.home_tab_me_title
            TitleType.Localized.LocalizedKey.Settings -> Res.string.settings_title
            TitleType.Localized.LocalizedKey.MastodonLocal -> Res.string.mastodon_tab_local_title
            TitleType.Localized.LocalizedKey.MastodonPublic -> Res.string.mastodon_tab_public_title
            TitleType.Localized.LocalizedKey.Featured -> Res.string.home_tab_featured_title
            TitleType.Localized.LocalizedKey.Bookmark -> Res.string.home_tab_bookmarks_title
            TitleType.Localized.LocalizedKey.Favourite -> Res.string.home_tab_favorite_title
            TitleType.Localized.LocalizedKey.List -> Res.string.home_tab_list_title
            TitleType.Localized.LocalizedKey.Feeds -> Res.string.home_tab_feeds_title
            TitleType.Localized.LocalizedKey.DirectMessage -> Res.string.dm_list_title
            TitleType.Localized.LocalizedKey.Rss -> Res.string.rss_title
            TitleType.Localized.LocalizedKey.Social -> Res.string.social_title
            TitleType.Localized.LocalizedKey.Antenna -> Res.string.antenna_title
            TitleType.Localized.LocalizedKey.MixedTimeline -> Res.string.mixed_timeline_title
            TitleType.Localized.LocalizedKey.Liked -> Res.string.liked_title
            TitleType.Localized.LocalizedKey.AllRssFeeds -> Res.string.all_rss_feeds_title
            TitleType.Localized.LocalizedKey.Posts -> Res.string.posts_title
            TitleType.Localized.LocalizedKey.Channel -> Res.string.channel_title
}
